[Certora] Documentation (#890)
morpho-org/vault-v2: certora/README.md

Reviewer: @bhargavbh maybe we could have something more minimal like [in midnight](https://github.com/morpho-org/midnight/tree/main/certora)?

This folder contains the formal verification of Vault V2 using CVL, Certora's Verification Language.

Vault V2 is a non-custodial ERC-4626 vault that allocates assets across markets through adapters. See the repository [`README`](../README.md) and [`src/VaultV2.sol`](../src/VaultV2.sol) for the protocol itself. The verified properties are listed below by theme, followed by the verification setup.

# Verified properties

## Core state and accounting

Global invariants and the accounting effects of each entry point.

> * [`Invariants.spec`](specs/Invariants.spec) collects the core state invariants. Performance and management fees, force-deallocation penalties, relative caps, and the maximum rate stay within their bounds; a non-zero fee always has a recipient; and the zero address has no shares. Total supply equals the sum of all balances, every allocation fits in an `int256`, virtual shares stay within their configured range, and the timelock for `decreaseTimelock` has the required value. The adapter list and `isAdapter` mapping remain mutually consistent, list entries are distinct, and, assuming an add-only registry, every adapter remains in the configured registry.

Reviewer: ```suggestion
* [`Invariants.spec`](specs/Invariants.spec) collects the core state invariants. Variables stay within their bounds; a non-zero fee always has a recipient; and the zero address has no shares. Total supply equals the sum of all balances, every allocation fits in an `int256`, virtual shares stay within their configured range. The adapter list and `isAdapter` mapping remain mutually consistent, list entries are distinct, and, assuming an add-only registry, every adapter remains in the configured registry.
```

> * [`TotalAssetsChange.spec`](specs/TotalAssetsChange.spec) pins down changes to `_totalAssets` when no interest accrues. Deposits and mints add exactly the transferred or previewed assets, withdrawals and redemptions subtract exactly the withdrawn or previewed assets, and `forceDeallocate` subtracts exactly its rounded-up penalty. Every other non-view entry point leaves `_totalAssets` unchanged.

Reviewer: ```suggestion
* [`TotalAssetsChange.spec`](specs/TotalAssetsChange.spec) pins down changes to `_totalAssets` when no interest accrues. Deposits and mints add exactly the transferred or previewed assets, withdrawals and redemptions subtract exactly the withdrawn or previewed assets, and `forceDeallocate` subtracts exactly its rounded-up penalty. Every other entry point leaves `_totalAssets` unchanged.
```

Author: true, view-functions are trivially covered

* [`TotalAssetsIsUpToDate.spec`](specs/TotalAssetsIsUpToDate.spec) checks that every state-changing entry point other than `accrueInterest` updates `_totalAssets` before reading it.
* [`AllocationVaultV2.spec`](specs/AllocationVaultV2.spec) checks that allocations change only through ERC-4626 entry and exit functions, `allocate`, `deallocate`, or `forceDeallocate`. An ERC-4626 operation can change allocations only when a liquidity adapter is set.
* [`AllocationsHierarchy.spec`](specs/AllocationsHierarchy.spec) checks the leaf-group structure used by adapter ids. A group's allocation always equals the sum of its leaf allocations and is therefore at least every individual leaf allocation.

## Shares and ERC-4626 behavior

Share-price movement, rounding, previews, and equivalent entry points.

> * [`ExchangeRate.spec`](specs/ExchangeRate.spec) checks that the seeded vault's share price does not decrease when the management fee is zero and interest has already been accrued. It separately shows that realizing a loss does not increase the share price and that deposit, mint, withdraw, and redeem use the tight protocol-favoring rounding direction.

Reviewer: ```suggestion
* [`ExchangeRate.spec`](specs/ExchangeRate.spec) checks that the vault's share price does not decrease when the management fee is zero and interest has already been accrued. It separately shows that realizing a loss does not increase the share price and that deposit, mint, withdraw, and redeem use the tightest rounding.
```

Author: using seeded is more accurate because we assume totalSupplyAssets>0 in the rule.

* [`PreviewFunctions.spec`](specs/PreviewFunctions.spec) checks that `previewDeposit`, `previewMint`, `previewWithdraw`, and `previewRedeem` return exactly the value produced by the corresponding successful ERC-4626 operation. A preview never reverts when its corresponding operation can succeed.
* [`EntrypointEquivalence.spec`](specs/EntrypointEquivalence.spec) checks that deposit and mint pass identical values to the internal entry path when their inputs and outputs match. It establishes the analogous equivalence between withdraw and redeem for the internal exit path.
* [`RoundTrip.spec`](specs/RoundTrip.spec) checks the rounding inequalities between asset/share conversions and every relevant composition of deposit, mint, withdraw, and redeem previews. These round trips cannot create value through inconsistent rounding.

## Adapter ids and allocation tracking

Adapters report stable risk ids and keep the vault's allocation accounting aligned with their underlying positions.

> * [`IdsMorphoMarketV1AdapterV2.spec`](specs/IdsMorphoMarketV1AdapterV2.spec) checks that the Morpho Market V1 adapter returns three deterministic and pairwise-distinct ids for a given market, including the id derived from the adapter address. The ids returned by `allocate` and `deallocate` match the adapter's reference list.

Reviewer: ```suggestion
* [`IdsMorphoMarketV1AdapterV2.spec`](specs/IdsMorphoMarketV1AdapterV2.spec) checks that the Morpho Market V1 adapter returns the three expected ids.
```

* [`AllocationMorphoVaultV1Adapter.spec`](specs/AllocationMorphoVaultV1Adapter.spec) checks the equivalent allocation updates for the Morpho Vault V1 adapter. After allocation or deallocation, the reported allocation equals the assets previewed from the adapter's MetaMorpho shares.
* [`ChangesMorphoMarketV1AdapterV2.spec`](specs/ChangesMorphoMarketV1AdapterV2.spec) and [`ChangesMorphoVaultV1Adapter.spec`](specs/ChangesMorphoVaultV1Adapter.spec) check each adapter's returned allocation change. Allocating and deallocating zero assets report the same change from the same state, and no reported change can make the current allocation negative.
* [`MarketIds.spec`](specs/MarketIds.spec) checks the Morpho Market V1 adapter's active-market list. Its entries are distinct, and a market with zero allocation is absent from the list.

## Caps and configuration delays

Risk limits and timelocked configuration cannot be bypassed.

> * [`RelativeCaps.spec`](specs/RelativeCaps.spec) checks that state-changing functions preserve every relative cap except the operations that may legitimately move outside it: interest or loss accrual, exits, cap decreases, and deallocation. The latter can increase a recorded allocation while realizing interest, whereas allocation would reject the same cap excess.

Reviewer: ```suggestion
* [`RelativeCaps.spec`](specs/RelativeCaps.spec) checks that relative caps are preserved, except by operations that may legitimately move outside it: interest or loss accrual, exits, cap decreases, and deallocation.
```

* [`EarliestTime.spec`](specs/EarliestTime.spec) checks the three ways a timelocked call can become executable: an existing submission, a fresh submission plus the current timelock, or a pending timelock decrease plus the new delay. For the covered configuration calls, the earliest execution time cannot move backward and the call cannot succeed before it.
* [`AbdicatedFunctions.spec`](specs/AbdicatedFunctions.spec) checks that an abdicated timelocked function cannot be called, that abdication is permanent, and that each affected configuration value remains unchanged in the direction the abdicated function would have modified it.
* [`Immutability.spec`](specs/Immutability.spec) checks that every `DELEGATECALL` targets the vault itself, preventing delegation to arbitrary implementation code.

## Authorization, input validation, and liveness

Privileged actions enforce their roles, while authorized accounts retain the operations needed to operate or derisk the vault.

> * [`Reverts.spec`](specs/Reverts.spec) checks the revert conditions or required input validation for timelocked configuration, ownership and metadata changes, submission and revocation, cap decreases, `forceDeallocate`, liquidity-adapter and max-rate changes, and share transfers. This covers role checks, timelocks and abdication, gates, allowances, balances, bounds, and non-payable calls.

Reviewer: ```suggestion
* [`Reverts.spec`](specs/Reverts.spec) checks the revert conditions or required input validation for timelocked configuration, ownership and metadata changes, submission and revocation, cap decreases, `forceDeallocate`, liquidity-adapter and max-rate changes, and share transfers.
```

* [`AllocateDeallocateReverts.spec`](specs/AllocateDeallocateReverts.spec) checks the exact authorization-level revert conditions for `allocate` and `deallocate`, assuming the adapter returns unique ids and allocation changes that satisfy the required cap and integer bounds. Allocation requires an allocator and a registered adapter; deallocation accepts an allocator or sentinel and also requires a registered adapter.
* [`AllocateDeallocateInputValidation.spec`](specs/AllocateDeallocateInputValidation.spec) checks that allocation rejects any returned id with a zero absolute cap and deallocation rejects any returned id with a zero recorded allocation, preventing interaction with unknown markets.

> * [`AccrueInterestReverts.spec`](specs/AccrueInterestReverts.spec) gives sufficient conditions under which `accrueInterestView` and `accrueInterest` do not revert. It bounds the returned total assets and fee shares and checks that a zero performance or management fee produces zero corresponding fee shares.

Reviewer: ```suggestion
* [`AccrueInterestReverts.spec`](specs/AccrueInterestReverts.spec) gives sufficient conditions under which `accrueInterestView` and `accrueInterest` do not revert. It bounds the returned total assets and fee shares as well.
```

* [`TokensNoAdapter.spec`](specs/TokensNoAdapter.spec) checks exact sender, receiver, and vault asset-balance changes on deposit and withdrawal when no liquidity adapter is configured.
* [`TokensMorphoMarketV1AdapterV2.spec`](specs/TokensMorphoMarketV1AdapterV2.spec) checks the same flows through a Morpho Market V1 liquidity adapter. Deposits move assets from the sender into Morpho without leaving balances on the vault or adapter; withdrawals consume idle vault assets first, then Morpho liquidity, and pay the receiver exactly.
* [`TokensMorphoVaultV1Adapter.spec`](specs/TokensMorphoVaultV1Adapter.spec) checks those token flows through a Morpho Vault V1 liquidity adapter and its underlying Morpho markets.
* [`SkimMorphoMarketV1AdapterV2.spec`](specs/SkimMorphoMarketV1AdapterV2.spec) checks that `skim` transfers only tokens already held by the Morpho Market V1 adapter and does not change its reported assets. It also checks that changing the skim recipient follows the adapter's timelock and abdication conditions.

> * [`SkimMorphoVaultV1Adapter.spec`](specs/SkimMorphoVaultV1Adapter.spec) checks the corresponding skim accounting for the Morpho Vault V1 adapter and requires the vault owner, with no ETH value, to change the skim recipient.

Reviewer: ```suggestion
* [`SkimMorphoVaultV1Adapter.spec`](specs/SkimMorphoVaultV1Adapter.spec) checks the corresponding skim accounting for the Morpho Vault V1 adapter and requires the vault owner to change the skim recipient.
```

## External calls and reentrancy

The vault does not expose an untrusted callback after entering an unsafe intermediate state.

> * [`Reentrancy.spec`](specs/Reentrancy.spec) checks that state-changing entry points make no external calls outside the vault itself, registered supported adapters, the asset token, Morpho Market V1, and MetaMorpho V1. The token and underlying markets are modeled as trusted not to reenter.

Reviewer: ```suggestion
* [`Reentrancy.spec`](specs/Reentrancy.spec) checks that entry points make no external calls outside the vault itself, registered supported adapters, the asset token, Morpho Market V1, and MetaMorpho V1, ensuring no reentrancy if the token and underlying markets are trusted not to reenter.
```

> * [`ReentrancyView.spec`](specs/ReentrancyView.spec) checks read-only reentrancy ordering: after an external static call follows a storage write, the vault performs no later storage write. Calls to the asset's `balanceOf`, adapters' `realAssets`, gate checks, and the adapter registry are the explicitly modeled view dependencies. The rule excludes `forceDeallocate`, which composes the separately analyzed `deallocate` and `withdraw` paths.

Reviewer: ```suggestion
* [`ReentrancyView.spec`](specs/ReentrancyView.spec) checks read-only reentrancy ordering: after an external static call follows a storage write, the vault performs no later storage write. Calls to the asset's `balanceOf`, adapters' `realAssets`, gate checks, and the adapter registry are assumed non-reentrants. The rule excludes `forceDeallocate`, which composes the separately analyzed `deallocate` and `withdraw` paths.
```

Author: resolved with equivalent simpler wording

* the reentrancy proofs trust the configured ERC-20 token, Morpho Market V1, and MetaMorpho V1 not to reenter the vault. Adapter calls are restricted to registered instances of the two verified adapter implementations;
* both rules in [`EarliestTime.spec`](specs/EarliestTime.spec) exclude `decreaseTimelock` because of a prover limitation around `msg.data`.

The [`confs`](confs) folder contains one configuration for every specification. Shared CVL utilities and Solidity harnesses are in [`helpers`](helpers).

# Getting started

Install the `certora-cli` package with `pip install certora-cli`. To verify a spec, pass its configuration file in [`certora/confs`](confs) to `certoraRun`. This requires a valid Certora key in the `CERTORAKEY` environment variable. The complete suite uses `solc-0.8.19`, `solc-0.8.21`, `solc-0.8.26`, and `solc-0.8.28`; the compiler versions required by a particular job are listed in its configuration.

Additional arguments can select a specific rule. For example, from the repository root:

```sh
certoraRun certora/confs/Invariants.conf --rule totalSupplyIsSumOfBalances
```

# Acknowledgments

> Some rules and invariants are derived from work by ChainSecurity during its audit of this repository.

Reviewer: ```suggestion
Some rules and invariants are derived from work by ChainSecurity during its VaultV2 audit.
```
